File: src/codegen/expander.py
from __future__ import annotations

import tempfile
import warnings
from dataclasses import dataclass, field
from pathlib import Path
from typing import Mapping

from codegen.config import Config, merge_pragma
from codegen.env import RunContext, build_env
from codegen.errors import AbortAll, BlockFailure
from codegen.executor import run_block as _run_block
from codegen.indent import apply_indent
from codegen.parser import Block, find_top_level_blocks
from codegen.scope import ScopeStore


@dataclass
class ExpandResult:
    text: str          # replacement text (no markers)
    ok: bool = True


def _splice(region: str, block: Block, replacement: str) -> str:
    """Replace the lines occupied by *block* in *region* with *replacement*."""
    lines = region.splitlines(keepends=True)
    before = lines[: block.start_line - 1]
    after = lines[block.end_line :]
    return "".join(before) + replacement + "".join(after)


def _write_tmp(content: str) -> Path:
    with tempfile.NamedTemporaryFile(
        mode="w", suffix=".txt", delete=False, encoding="utf-8"
    ) as f:
        f.write(content)
        return Path(f.name)


def _keep_as_comment(original_raw: str, block: Block, output: str) -> str:
    """Prepend the original block as a comment above the generated output (§5)."""
    cs = block.comment_syntax
    lines = original_raw.splitlines(keepends=True)
    if cs.is_block:
        commented = (
            f"{block.indent}// === codegen source (kept) ===\n"
            + "".join(f"{block.indent}// {line.rstrip()}\n" for line in lines)
            + f"{block.indent}// === end ===\n"
        )
    else:
        commented = (
            f"{block.indent}{cs.open} === codegen source (kept) ===\n"
            + "".join(f"{block.indent}{cs.open} {line.rstrip()}\n" for line in lines)
            + f"{block.indent}{cs.open} === end ===\n"
        )
    return commented + output


def expand_block(
    block: Block,
    parent_cfg: Config,
    scope: ScopeStore,
    ctx: RunContext,
) -> ExpandResult:
    """Expand a single top-level block through up to max_passes rounds (§6.2).

    Returns ExpandResult with the replacement text (no markers, may contain
    generated code plus any keep_as_comment header).

    Raises BlockFailure on error (caller decides on_error behaviour).
    """
    block_cfg = merge_pragma(parent_cfg, block.pragma, source=f"{block.file_path}:{block.start_line}")

    # Snapshot scope dicts before touching this block (§10.5)
    scope.open_block()
    scope.snapshot()
    snapshot_finalized = False

    # Create origin_block snapshot (§4.2)
    origin_block_path = _write_tmp(block.inner_text)

    try:
        # region starts as the full raw block text (including markers).
        # Each pass replaces block(s) found in region with their stdout.
        region = block.raw_block_text
        elapsed_total = 0.0
        pass_outputs: list[str] = []
        pass_idx = 0
        original_raw = block.raw_block_text

        while pass_idx < block_cfg.max_passes:
            inner_blocks = find_top_level_blocks(
                region,
                file_path=block.file_path,
                markers=(block_cfg.markers[0], block_cfg.markers[1]),
                cs=block.comment_syntax,
            )
            if not inner_blocks:
                break  # stable

            for ib in inner_blocks:
                # For nested sub-blocks, origin_block is the sub-block's inner text
                ib_origin = _write_tmp(ib.inner_text)
                ib_env = build_env(
                    block_cfg.extra_env,
                    scope,
                    RunContext(
                        invoke_cwd=ctx.invoke_cwd,
                        targets=ctx.targets,
                        file_path=ctx.file_path,
                        origin_file_path=ctx.origin_file_path,
                        origin_block_path=ib_origin,
                    ),
                )

                cwd = block_cfg.cwd or ctx.invoke_cwd
                stdout, elapsed = _run_block(
                    ib,
                    env=ib_env,
                    cwd=cwd,
                    max_pass_time=block_cfg.max_pass_time,
                    pass_outputs=pass_outputs,
                )
                try:
                    ib_origin.unlink()
                except OSError:
                    pass

                pass_outputs.append(stdout)
                elapsed_total += elapsed

                if elapsed_total > block_cfg.max_total_time:
                    raise BlockFailure(
                        block=ib,
                        reason="timeout:total",
                        pass_outputs=list(pass_outputs),
                    )

                indented = apply_indent(stdout, ib.indent, block_cfg.auto_indent)
                region = _splice(region, ib, indented)

            pass_idx += 1

        else:
            # Loop exhausted without break — warn only if blocks remain in region
            remaining = find_top_level_blocks(
                region,
                file_path=block.file_path,
                markers=(block_cfg.markers[0], block_cfg.markers[1]),
                cs=block.comment_syntax,
            )
            if remaining:
                warnings.warn(
                    f"{block.file_path}:{block.start_line}: max_passes={block_cfg.max_passes} reached; "
                    "expansion may be incomplete",
                    stacklevel=2,
                )

        scope.commit()
        snapshot_finalized = True

        if block_cfg.keep_as_comment:
            region = _keep_as_comment(original_raw, block, region)

        return ExpandResult(text=region, ok=True)

    except BlockFailure:
        scope.restore()
        snapshot_finalized = True
        raise
    finally:
        if not snapshot_finalized:
            # Non-BlockFailure exception: restore so the snapshot doesn't leak.
            scope.restore()
        try:
            origin_block_path.unlink()
        except OSError:
            pass
        scope.close_block()


def process_content(
    content: str,
    cfg: Config,
    scope: ScopeStore,
    ctx: RunContext,
) -> tuple[str, bool]:
    """Process all top-level blocks in *content*.

    Returns (updated_text, had_failure). Each block is expanded in place;
    on_error decides whether to keep going, abort the file, or raise AbortAll.
    Per §10.4, the failure diagnostic is always emitted to stderr regardless
    of which on_error mode is active.
    """
    from codegen.comment_syntax import lookup

    cs = lookup(ctx.file_path, overrides=dict(cfg.comment_syntax_overrides))
    if cs is None:
        return content, False  # unknown extension: nothing to do

    blocks = find_top_level_blocks(
        content,
        file_path=ctx.file_path,
        markers=cfg.markers,
        cs=cs,
    )
    if not blocks:
        return content, False

    # Work with a mutable list of lines; offsets shift as we replace blocks.
    # Easiest: rebuild content string after each block (blocks are sequential).
    result = content
    line_offset = 0  # track how many lines the content has shifted so far
    had_failure = False

    for block in blocks:
        # Adjust block position by accumulated line offset
        adjusted = Block(
            file_path=block.file_path,
            start_line=block.start_line + line_offset,
            end_line=block.end_line + line_offset,
            indent=block.indent,
            comment_syntax=block.comment_syntax,
            raw_block_text=block.raw_block_text,
            inner_text=block.inner_text,
            shebang=block.shebang,
            pragma=block.pragma,
            body=block.body,
        )

        try:
            expand_res = expand_block(adjusted, cfg, scope, ctx)
        except BlockFailure as exc:
            _emit_failure(exc)  # §10.4: always emit diagnostic
            had_failure = True
            if cfg.on_error == "abort_all":
                raise AbortAll(exc) from exc
            if cfg.on_error == "abort_file":
                raise
            # continue: keep original block text, move on
            continue

        replacement = expand_res.text
        # If the original block was line-terminated, the replacement must be too —
        # otherwise the line that follows the block would be merged into the last
        # line of the replacement, corrupting the file and shifting later blocks.
        if adjusted.raw_block_text.endswith("\n") and replacement and not replacement.endswith("\n"):
            replacement += "\n"

        old_line_count = len(adjusted.raw_block_text.splitlines())
        new_line_count = len(replacement.splitlines())
        line_offset += new_line_count - old_line_count

        result = _splice(result, adjusted, replacement)

    return result, had_failure


def _emit_failure(exc: BlockFailure) -> None:
    import sys

    b = exc.block
    print(f"codegen: block 失敗 — {b.file_path} 行 {b.start_line}", file=sys.stderr)
    print("原始 block:", file=sys.stderr)
    for line in b.inner_text.splitlines():
        print(f"  {line}", file=sys.stderr)
    for i, out in enumerate(exc.pass_outputs, 1):
        print(f"pass {i} stdout:", file=sys.stderr)
        for line in out.splitlines():
            print(f"  {line}", file=sys.stderr)
    print(f"失敗原因：{exc.reason}", file=sys.stderr)
    if exc.last_stderr:
        print("stderr:", file=sys.stderr)
        for line in exc.last_stderr.splitlines():
            print(f"  {line}", file=sys.stderr)

File: cpp/src/expander.cpp
#include "expander.hpp"
#include "executor.hpp"
#include "indent.hpp"
#include "utils.hpp"

#include <fstream>
#include <iostream>
#include <sstream>

#ifdef _WIN32
#include <windows.h>
static fs::path write_tmp_file(const std::string& content, bool /*world_readable*/ = false) {
    char tmp[MAX_PATH];
    char dir[MAX_PATH];
    GetTempPathA(MAX_PATH, dir);
    GetTempFileNameA(dir, "cg", 0, tmp);
    write_file(fs::path(tmp), content);
    return tmp;
}
#else
#include <sys/stat.h>
#include <unistd.h>
// world_readable: a dropped-privilege block (run_as_user) reads these via
// CODEGEN_ORIGIN_* and would otherwise be denied by the 0600 default.
static fs::path write_tmp_file(const std::string& content, bool world_readable = false) {
    char tmpl[] = "/tmp/codegen_tmp_XXXXXX";
    int fd = mkstemp(tmpl);
    if (fd < 0) throw std::runtime_error("mkstemp failed");
    if (world_readable) fchmod(fd, 0644);
    if (write(fd, content.data(), content.size()) < 0) {
        close(fd);
        throw std::runtime_error("write tmp failed");
    }
    close(fd);
    return tmpl;
}
#endif

// Replace lines [start_line-1 .. end_line-1] (0-indexed) with replacement.
static std::string splice(const std::string& region,
                           int start_line, int end_line,
                           const std::string& replacement) {
    auto lines = splitlines(region, true);
    std::string result;
    for (int i = 0; i < (int)lines.size(); i++) {
        if (i < start_line - 1 || i >= end_line)
            result += lines[i];
        else if (i == start_line - 1)
            result += replacement;
    }
    return result;
}

static std::string keep_as_comment(const std::string& original_raw,
                                    const Block& block,
                                    const std::string& output) {
    const CommentSyntax& cs = block.comment_syntax;
    auto lines = splitlines(original_raw, false);
    std::string commented;
    if (cs.is_block()) {
        commented += block.indent + "// === codegen source (kept) ===\n";
        for (auto& l : lines)
            commented += block.indent + "// " + l + "\n";
        commented += block.indent + "// === end ===\n";
    } else {
        commented += block.indent + cs.open + " === codegen source (kept) ===\n";
        for (auto& l : lines)
            commented += block.indent + cs.open + " " + l + "\n";
        commented += block.indent + cs.open + " === end ===\n";
    }
    return commented + output;
}

std::string expand_block(
    const Block& block,
    const Config& parent_cfg,
    ScopeStore& scope,
    const RunContext& ctx)
{
    Config block_cfg = merge_from_strings(
        parent_cfg, block.pragma, /*is_pragma=*/true);

    scope.open_block();
    scope.snapshot();
    bool snapshot_done = false;

    const bool world = block_cfg.run_as_user.has_value();
    fs::path origin_block_path = write_tmp_file(block.inner_text, world);

    try {
        std::string region = block.raw_block_text;
        double elapsed_total = 0.0;
        std::vector<std::string> pass_outputs;
        int pass_idx = 0;
        std::string original_raw = block.raw_block_text;

        while (pass_idx < block_cfg.max_passes) {
            auto inner_blocks = find_top_level_blocks(
                region, block.file_path,
                {block_cfg.markers.first, block_cfg.markers.second},
                block.comment_syntax);

            if (inner_blocks.empty()) break; // stable

            for (auto& ib : inner_blocks) {
                fs::path ib_origin = write_tmp_file(ib.inner_text, world);

                RunContext ib_ctx{
                    ctx.invoke_cwd,
                    ctx.targets,
                    ctx.file_path,
                    ctx.origin_file_path,
                    ib_origin,
                };

                auto ib_env = build_env(block_cfg.extra_env, scope, ib_ctx);
                fs::path eff_cwd = block_cfg.cwd ? *block_cfg.cwd : ctx.invoke_cwd;

                auto [stdout_str, elapsed] = run_block(
                    ib, ib_env, eff_cwd,
                    block_cfg.max_pass_time,
                    pass_outputs,
                    block_cfg.run_as_user);

                std::error_code ec;
                fs::remove(ib_origin, ec);

                pass_outputs.push_back(stdout_str);
                elapsed_total += elapsed;

                if (elapsed_total > block_cfg.max_total_time) {
                    throw BlockFailure(ib, "timeout:total", pass_outputs);
                }

                std::string indented = apply_indent(
                    stdout_str, ib.indent, block_cfg.auto_indent);
                region = splice(region,
                                ib.start_line, ib.end_line,
                                indented);
            }
            pass_idx++;
        }

        // Check for remaining unexpanded blocks (max_passes exceeded)
        if (pass_idx == block_cfg.max_passes) {
            auto remaining = find_top_level_blocks(
                region, block.file_path,
                {block_cfg.markers.first, block_cfg.markers.second},
                block.comment_syntax);
            if (!remaining.empty()) {
                std::cerr << block.file_path.string() << ":" << block.start_line
                          << ": warning: max_passes=" << block_cfg.max_passes
                          << " reached; expansion may be incomplete\n";
            }
        }

        scope.commit();
        snapshot_done = true;

        if (block_cfg.keep_as_comment)
            region = keep_as_comment(original_raw, block, region);

        std::error_code ec;
        fs::remove(origin_block_path, ec);
        scope.close_block();
        return region;

    } catch (BlockFailure&) {
        scope.restore();
        snapshot_done = true;
        std::error_code ec;
        fs::remove(origin_block_path, ec);
        scope.close_block();
        throw;
    } catch (...) {
        if (!snapshot_done) scope.restore();
        std::error_code ec;
        fs::remove(origin_block_path, ec);
        scope.close_block();
        throw;
    }
}

void emit_failure(const BlockFailure& exc) {
    const Block& b = exc.block;
    std::cerr << "codegen: block 失敗 — " << b.file_path.string()
              << " 行 " << b.start_line << "\n";
    std::cerr << "原始 block:\n";
    for (auto& line : splitlines(b.inner_text))
        std::cerr << "  " << line << "\n";
    for (size_t i = 0; i < exc.pass_outputs.size(); i++) {
        std::cerr << "pass " << (i + 1) << " stdout:\n";
        for (auto& line : splitlines(exc.pass_outputs[i]))
            std::cerr << "  " << line << "\n";
    }
    std::cerr << "失敗原因：" << exc.reason << "\n";
    if (!exc.last_stderr.empty()) {
        std::cerr << "stderr:\n";
        for (auto& line : splitlines(exc.last_stderr))
            std::cerr << "  " << line << "\n";
    }
}

std::pair<std::string, bool> process_content(
    const std::string& content,
    const Config& cfg,
    ScopeStore& scope,
    const RunContext& ctx)
{
    auto cs_opt = lookup_syntax(ctx.file_path, &cfg.comment_syntax_overrides);
    if (!cs_opt) return {content, false};

    auto blocks = find_top_level_blocks(
        content, ctx.file_path, cfg.markers, *cs_opt);
    if (blocks.empty()) return {content, false};

    std::string result = content;
    int line_offset = 0;
    bool had_failure = false;

    for (auto& block : blocks) {
        Block adjusted = block;
        adjusted.start_line += line_offset;
        adjusted.end_line   += line_offset;

        std::string replacement;
        try {
            replacement = expand_block(adjusted, cfg, scope, ctx);
        } catch (BlockFailure& exc) {
            emit_failure(exc);
            had_failure = true;
            if (cfg.on_error == "abort_all")
                throw AbortAll(std::move(exc));
            if (cfg.on_error == "abort_file")
                throw;
            // continue: leave original block text
            continue;
        }

        // Preserve trailing newline of original block
        if (ends_with(adjusted.raw_block_text, "\n") &&
            !replacement.empty() && !ends_with(replacement, "\n"))
            replacement += "\n";

        int old_lines = (int)splitlines(adjusted.raw_block_text).size();
        int new_lines = (int)splitlines(replacement).size();
        line_offset += new_lines - old_lines;

        result = splice(result,
                        adjusted.start_line, adjusted.end_line,
                        replacement);
    }

    return {result, had_failure};
}
